File: algorithms/medium/spiral_matrix_iv.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
#[derive(Clone, Copy)]
enum Direction {
    Up,
    Right,
    Down,
    Left,
}

use Direction::*;

fn turn(x: &mut usize, y: &mut usize, d: &mut Direction) {
    match d {
        Up => {*x += 1; *y += 1; *d = Right;},
        Down => {*x -= 1; *y -= 1; *d = Left;},
        Left => {*x -= 1; *y += 1; *d = Up;},
        Right => {*x += 1; *y -= 1; *d = Down;},
    };
}

fn next(x: &mut usize, y: &mut usize, d: Direction) {
    match d {
        Up => *x -= 1,
        Down => *x += 1,
        Left => *y -= 1,
        Right => *y += 1,
    };
}

impl Solution {
    pub fn spiral_matrix(m: i32, n: i32, head: Option<Box<ListNode>>) -> Vec<Vec<i32>> {
        let m = m as usize;
        let n = n as usize;
        let mut ans = vec![vec![-1; n]; m];
        let mut ct = 0;
        let mut i = 0;
        let mut j = 0;
        let mut d = Right;
        let mut curr = head;
        while curr.is_some() && ct < m * n {
            if i >= m || j >= n || ans[i][j] != -1 {
                turn(&mut i, &mut j, &mut d);
                continue;
            }
            if let Some(node) = curr {
                ans[i][j] = node.val;
                curr = node.next;
            }
            next(&mut i, &mut j, d);
            ct += 1;
        }
        ans
    }
}

File: algorithms/medium/spiral_matrix_iv.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec spiral_matrix(m :: integer, n :: integer, head :: ListNode.t() | nil) :: [[integer]]
  def spiral_matrix(m, n, head) do
    dt = aux(%{}, head, {0, 0, :right}, m, n)

    for i <- 0..(m - 1) do
      for j <- 0..(n - 1), do: dt[{i, j}] || -1
    end
  end

  def aux(acc, nil, _pos, _m, _n), do: acc

  def aux(acc, _head, _pos, m, n) when map_size(acc) == m * n do
    acc
  end

  def aux(acc, head, {x, y, dir}, m, n) do
    if x < 0 || x >= m || y < 0 || y >= n || acc[{x, y}] != nil do
      aux(acc, head, turn({x, y, dir}), m, n)
    else
      {val, head} = if head, do: {head.val, head.next}, else: {-1, nil}

      Map.put(acc, {x, y}, val) |> aux(head, next({x, y, dir}), m, n)
    end
  end

  def turn({x, y, dir}) do
    case dir do
      :up -> {x + 1, y + 1, :right}
      :down -> {x - 1, y - 1, :left}
      :left -> {x - 1, y + 1, :up}
      :right -> {x + 1, y - 1, :down}
    end
  end

  def next({x, y, dir}) do
    case dir do
      :up -> {x - 1, y, dir}
      :down -> {x + 1, y, dir}
      :left -> {x, y - 1, dir}
      :right -> {x, y + 1, dir}
    end
  end
end
